modules/network/src/main/java/org/apache/ignite/network/LazyStripedExecutor.java

```diff
@@ -0,0 +1,97 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.network;
+
+import java.util.Objects;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.stream.IntStream;
+import org.apache.ignite.internal.close.ManuallyCloseable;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.internal.util.IgniteUtils;
+
+/**
+ * Lazy striped executor indexed with short type. A thread is created on first execution with an index and remains active forever.
+ *
+ * <p>After having been stopped, it never executes anything.
+ */
+class LazyStripedExecutor implements ManuallyCloseable {
+    private static final IgniteLogger LOG = Loggers.forClass(LazyStripedExecutor.class);
+
+    private final String nodeName;
+    private final String poolName;
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final AtomicReferenceArray<ExecutorService> array = new AtomicReferenceArray<>(Short.MAX_VALUE + 1);
+
+    LazyStripedExecutor(String nodeName, String poolName) {
+        this.nodeName = nodeName;
+        this.poolName = poolName;
+    }
+
+    /**
+     * Executes a command on a stripe with the given index. If the executor is stopped, does nothing.
+     *
+     * @param index Index of the stripe.
+     */
+    public void execute(short index, Runnable command) {
+        assert index >= 0 : "Index is negative: " + index;
+
+        if (!busyLock.enterBusy()) {
+            return;
+        }
+
+        try {
+            executorFor(index).execute(command);
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    private Executor executorFor(short index) {
+        return array.updateAndGet(
+                index,
+                existing -> existing != null
+                        ? existing
+                        : Executors.newSingleThreadExecutor(NamedThreadFactory.create(nodeName, poolName + "-" + index, LOG))
```

Comment:
I see your intentions, but technically speaking, this code is incorrect. There's no guarantee that you'll only allocate a single pool, closure might be executed several times with `existing == null`.
Could you please re-write initialization using a good old double-checked locking instead?
Atomic reference array will stay, of course

Reply:
Nice catch, thank you! Fixed.